File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrack(index, cur, curSum):
            if curSum == target:
                ans.append(cur[:])
                return

            if curSum > target:
                return

            for i in range(index, len(candidates)):
                elem = candidates[i]
                if elem + curSum <= target:
                    cur.append(elem)
                    backtrack(i, cur, curSum + elem)
                    cur.pop()

        ans = []
        backtrack(0, [], 0)
        return ans

File: 39-combination-sum/combination-sum.php
class Solution {

	/**
	* @param Integer[] $candidates
	* @param Integer $target
	* @return Integer[][]
	*/
	function combinationSum($candidates, $target) {

		$backtrack = function($comb, $curSum, $index) use ($target, $candidates, &$ans, &$backtrack) {

			if ($curSum == $target) {
				$ans[] = $comb;
				return;
			}

			if ($curSum > $target || $index >= count($candidates)) {
				return;
			}

			for ($i = $index; $i < count($candidates); $i++) {

				if ($curSum + $candidates[$i] <= $target) {
					$comb[] = $candidates[$i];
					$backtrack($comb, $curSum+$candidates[$i], $i);
					array_pop($comb);
				}

			}
		};

		$ans = [];
		$backtrack([], 0, 0);
		return $ans;

	}
}